homeassistant/components/media_player/sonos.py

```diff
@@ -0,0 +1,194 @@
+    def name(self):
+        """ Returns the name of the device. """
+        return self._name
+
+    @property
+    def state(self):
+        """ Returns the state of the device. """
+        if self._status == 'PAUSED_PLAYBACK':
+            return STATE_PAUSED
+        if self._status == 'PLAYING':
+            return STATE_PLAYING
+        if self._status == 'STOPPED':
+            return STATE_IDLE
+        return STATE_UNKNOWN
+
+    def update(self):
+        """ Retrieve latest state. """
+        self._status = self._player.get_current_transport_info().get(
+            'current_transport_state')
+
+    @property
+    def volume_level(self):
+        """ Volume level of the media player (0..1). """
+        if 'mixer volume' in self._status:
+            return self._player.volume / 100.0
+
+    @property
+    def is_volume_muted(self):
+        return self._player.mute
+
+    @property
+    def media_content_id(self):
+        """ Content ID of current playing media. """
+        return self._trackinfo.get('title', None)
+
+    @property
+    def media_content_type(self):
+        """ Content type of current playing media. """
+        return MEDIA_TYPE_MUSIC
+
+    @property
+    def media_duration(self):
+        """ Duration of current playing media in seconds. """
+        dur = self._trackinfo.get('duration', '0:00')
+        return sum(60 ** x[0] * int(x[1]) for x in
+                   enumerate(reversed(dur.split(':'))))
+
+    @property
+    def media_image_url(self):
+        """ Image url of current playing media. """
+        if 'album_art' in self._trackinfo:
+            return self._trackinfo['album_art']
+
+    @property
+    def media_title(self):
+        """ Title of current playing media. """
+        if 'artist' in self._trackinfo and 'title' in self._trackinfo:
+            return '{artist} - {title}'.format(
+                artist=self._trackinfo['artist'],
+                title=self._trackinfo['title']
+            )
+        if 'title' in self._status:
+            return self._trackinfo['title']
+
+    @property
+    def supported_media_commands(self):
+        """ Flags of media commands that are supported. """
+        return SUPPORT_SONOS
+
+    def turn_off(self):
+        """ turn_off media player. """
+        self._player.pause()
+        self.update_ha_state()
+
+    def volume_up(self):
+        """ volume_up media player. """
+        self._player.volume += 1
+        self.update_ha_state()
+
+    def volume_down(self):
+        """ volume_down media player. """
+        self._player.volume -= 1
+        self.update_ha_state()
+
+    def set_volume_level(self, volume):
+        """ set volume level, range 0..1. """
+        self._player.volume = str(int(volume * 100))
+        self.update_ha_state()
+
+    def mute_volume(self, mute):
+        """ mute (true) or unmute (false) media player. """
+        self._player.mute = mute
+        self.update_ha_state()
+
+    def media_play_pause(self):
```

Comment:
This method is not needed, it is [implemented](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/media_player/__init__.py#L469-L474) like this in the base class.